consensus/snapshot/local_snapshots/local_snapshots_manager.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2019 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include "consensus/snapshot/local_snapshots/local_snapshots_manager.h"
+#include <inttypes.h>
+#include "common/storage/connection.h"
+#include "consensus/snapshot/local_snapshots/conf.h"
+#include "consensus/snapshot/snapshots_service.h"
+#include "utils/logger_helper.h"
+#include "utils/macros.h"
+#include "utils/time.h"
+
+#define LOCAL_SNAPSHOTS_RESCAN_INTERVAL_MS 10000
+
+#define LOCAL_SNAPSHOTS_MANAGER_LOGGER_ID "local_snapshots_manager"
+
+static logger_id_t logger_id;
+
+static void *local_snapshots_manager_routine(void *arg) {
+  local_snapshots_manager_t *lsm = (local_snapshots_manager_t *)arg;
+  lock_handle_t lock_cond;
+  retcode_t err;
+  size_t exponential_delay_factor = 1;
+  uint64_t start_timestamp, end_timestamp;
+
+  lock_handle_init(&lock_cond);
+  lock_handle_lock(&lock_cond);
+
+  tangle_t tangle;
+  connection_config_t db_conf = {.db_path = lsm->conf->db_path};
+  if ((err = iota_tangle_init(&tangle, &db_conf))) {
+    log_critical(logger_id, "Failed in initializing db\n");
+    return NULL;
+  }
+
+  bool skip_check = false;
+
+  while (lsm->running) {
+    if (skip_check || iota_local_snapshots_manager_should_take_snapshot(lsm, &tangle)) {
+      start_timestamp = current_timestamp_ms();
+      err = iota_snapshots_service_take_snapshot(lsm->snapshots_service, &tangle);
+      if (err == RC_OK) {
+        exponential_delay_factor = 1;
+        end_timestamp = current_timestamp_ms();
+        log_info(logger_id, "Local snapshot completed successfully in %" PRIu64 " ms\n",
+                 end_timestamp - start_timestamp);
+        if (iota_tangle_transaction_count(&tangle, &lsm->last_snapshot_transactions_count) != RC_OK) {
+          log_critical(logger_id, "Failed in querying db size\n");
+          goto cleanup;
+        }
+      } else {
+        exponential_delay_factor *= 2;
+        log_warning(logger_id, "Local snapshot is delayed in %d ms, error code: %d\n",
+                    exponential_delay_factor * LOCAL_SNAPSHOTS_RESCAN_INTERVAL_MS, err);
+        cond_handle_timedwait(&lsm->cond_local_snapshots, &lock_cond,
+                              exponential_delay_factor * LOCAL_SNAPSHOTS_RESCAN_INTERVAL_MS);
+      }
+    }
+    if (!iota_local_snapshots_manager_should_take_snapshot(lsm, &tangle)) {
+      cond_handle_timedwait(&lsm->cond_local_snapshots, &lock_cond, LOCAL_SNAPSHOTS_RESCAN_INTERVAL_MS);
+      skip_check = false;
+    } else {
+      skip_check = true;
+    }
+  }
```

Comment:
I don't think so, because sometimes we want to take several snapshots in a row, and for that with your suggestions i will have to set `exponential_delay_factor` to 0 so `cond_handle_timedwait` won't wait, i prefer the current